node/network/availability-distribution/src/error.rs

```diff
@@ -28,116 +28,88 @@ use polkadot_subsystem::{ChainApiError, SubsystemError};
 
 use crate::LOG_TARGET;
 
-#[derive(Debug, Error, derive_more::From)]
-#[error(transparent)]
+#[allow(missing_docs)]
+#[fatality::fatality(splitable)]
 pub enum Error {
-	/// All fatal errors.
-	Fatal(Fatal),
-	/// All nonfatal/potentially recoverable errors.
-	NonFatal(NonFatal),
-}
-
-impl From<runtime::Error> for Error {
-	fn from(o: runtime::Error) -> Self {
-		match o {
-			runtime::Error::Fatal(f) => Self::Fatal(Fatal::Runtime(f)),
-			runtime::Error::NonFatal(f) => Self::NonFatal(NonFatal::Runtime(f)),
-		}
-	}
-}
-
-/// Fatal errors of this subsystem.
-#[derive(Debug, Error)]
-pub enum Fatal {
-	/// Spawning a running task failed.
+	#[fatal]
 	#[error("Spawning subsystem task failed: {0}")]
 	SpawnTask(#[source] SubsystemError),
 
-	/// Requester stream exhausted.
+	#[fatal]
 	#[error("Erasure chunk requester stream exhausted")]
 	RequesterExhausted,
 
+	#[fatal]
 	#[error("Receive channel closed: {0}")]
 	IncomingMessageChannel(#[source] SubsystemError),
 
-	/// Errors coming from runtime::Runtime.
+	#[fatal(forward)]
 	#[error("Error while accessing runtime information: {0}")]
-	Runtime(#[from] runtime::Fatal),
+	Runtime(#[from] runtime::Error),
 
+	#[fatal]
 	#[error("Oneshot for receiving response from Chain API got cancelled")]
 	ChainApiSenderDropped(#[source] oneshot::Canceled),
 
+	#[fatal]
 	#[error("Retrieving response from Chain API unexpectedly failed with error: {0}")]
 	ChainApi(#[from] ChainApiError),
-}
 
-/// Non-fatal errors of this subsystem.
-#[derive(Debug, Error)]
-pub enum NonFatal {
-	/// av-store will drop the sender on any error that happens.
+	// av-store will drop the sender on any error that happens.
 	#[error("Response channel to obtain chunk failed")]
 	QueryChunkResponseChannel(#[source] oneshot::Canceled),
 
-	/// av-store will drop the sender on any error that happens.
+	// av-store will drop the sender on any error that happens.
 	#[error("Response channel to obtain available data failed")]
 	QueryAvailableDataResponseChannel(#[source] oneshot::Canceled),
 
-	/// We tried accessing a session that was not cached.
+	// We tried accessing a session that was not cached.
+	#[fatal(false)]
```

Comment:
most non-fatal errors don't have `fatal(false)` but this does. Why?

Reply:
I think this was the earliest implementation of fatality, it still does what's in the tin, not having any annotations has the same effect.

I will remove it and do a normalization pass

Reply:
Strike that. It's a remainder from debugging a `fatality` bug. Fixed in [a7b9379](https://github.com/paritytech/polkadot/pull/4448/commits/a7b9379deced3f324135f9472ad260af60c7bf50)